### ISSUE TITLE
unnecessary_string_escapes flags strings where a backslash terminates a variable interpolation

### ISSUE THREAD

**Issue by srawlins** [dart-lang/linter#3498]
**Describe the issue**

unnecessary_string_escapes flags strings where a backslash terminates a variable interpolation

**To Reproduce**

```dart
void main() {
  String hello = 'hello';
  print('Sort$hello\Numbers');
}
```

The linter flags this with `unnecessary_string_escapes` and `dart fix` will remove the backslash, creating a compile-time error.

**Expected behavior**

The lint rule should not be reported.

**Comment by bwilkerson** [comment]
Is using a backslash the encouraged style for dealing with this situation? The alternative I can think of is braces (`'Sort${hello}Numbers'`), but I don't know which is preferred (but I seem to recall that we special cased the `unnecessary_braces_in_interpolation` lint to not report in those cases).

If braces are preferred, then we might want to change the fix for escapes to add the braces when necessary instead of just not reporting it. If an escape is preferred then we might want to update the other lint and it's fix. And if there's no preference, then probably both lints should be silent when removing the characters would create an error.

**Comment by srawlins** [comment]
The braces are definitely more idiomatic than the backslash.

> And if there's no preference, then probably both lints should be silent when removing the characters would create an error.

This would be my preference. If removing an escape would change the meaning of the string (and break the library), then it's hard to call it "unnecessary."

**Comment by lrhn** [comment]
I think we should really push back against using escapes as a way to avoid braces. It's error-prone.

Say you do `"Answer$separator\yes"` and decide to change the answer to "no", so you write `"Answer$separator\no"`.
Then you have a newline in your string.

I wouldn't allow unnecessary escapes (defined as escapes where `\<char>` means the same as `<char>`). If removing the escape means inserting braces, than we should do that in the fix, but it's not a good reason to not report the unnecessary escape.

(I really wish we had never allowed unnecessary escapes at all. They add nothing but noise if used, and it prevents adding new escapes sequences in the future. Consider this as my intent to remove them - whether I'll ever succeed might be up to this lint :wink:).
